internal/settings/providers/gcp/provider.go

```diff
@@ -0,0 +1,183 @@
+package gcp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/qdm12/ddns-updater/internal/models"
+	"github.com/qdm12/ddns-updater/internal/settings/constants"
+	"github.com/qdm12/ddns-updater/internal/settings/errors"
+	"github.com/qdm12/ddns-updater/internal/settings/utils"
+	"github.com/qdm12/ddns-updater/pkg/publicip/ipversion"
+	"golang.org/x/exp/slices"
+	clouddns "google.golang.org/api/dns/v1"
+	"google.golang.org/api/googleapi"
+	"net"
+	"net/http"
+	"os"
+)
+
+type provider struct {
+	domain    string
+	host      string
+	zone      string
+	project   string
+	ipVersion ipversion.IPVersion
+}
+
+func New(data json.RawMessage, domain, host string,
+	ipVersion ipversion.IPVersion) (p *provider, err error) {
+	extraSettings := struct {
+		Project string `json:"project"`
+		Zone    string `json:"zone"`
+	}{}
+	if err := json.Unmarshal(data, &extraSettings); err != nil {
+		return nil, err
+	}
+	p = &provider{
+		domain:    domain,
+		host:      host,
+		ipVersion: ipVersion,
+		zone:      extraSettings.Zone,
+		project:   extraSettings.Project,
+	}
+	if err := p.isValid(); err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func (p *provider) isValid() error {
+	_, present := os.LookupEnv("GOOGLE_APPLICATION_CREDENTIALS")
+	if !present {
+		return errors.ErrEmptyCredentials
+	}
+
+	if len(p.project) == 0 {
+		return errors.ErrEmptyProject
+	}
+	if len(p.zone) == 0 {
+		return errors.ErrEmptyZoneIdentifier
+	}
+	return nil
+}
+
+func (p *provider) String() string {
+	return utils.ToString(p.domain, p.host, constants.GCP, p.ipVersion)
+}
+
+func (p *provider) Domain() string {
+	return p.domain
+}
+
+func (p *provider) Host() string {
+	return p.host
+}
+
+func (p *provider) IPVersion() ipversion.IPVersion {
+	return p.ipVersion
+}
+
+func (p *provider) Proxied() bool {
+	return false
+}
+
+func (p *provider) BuildDomainName() string {
+	return utils.BuildDomainName(p.host, p.domain)
+}
+
+func (p *provider) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain:    models.HTML(fmt.Sprintf("<a href=\"http://%s\">%s</a>", p.BuildDomainName(), p.BuildDomainName())),
+		Host:      models.HTML(p.Host()),
+		Provider:  "<a href=\"https://cloud.google.com/\">Google Cloud</a>",
+		IPVersion: models.HTML(p.ipVersion.String()),
+	}
+}
+
+func (p *provider) Update(ctx context.Context, _ *http.Client, ip net.IP) (newIP net.IP, err error) {
+	recordType := constants.A
+	if ip.To4() == nil {
+		recordType = constants.AAAA
+	}
+
+	dnsClient, err := clouddns.NewService(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	record, err := p.getRecord(dnsClient, recordType)
+	if err != nil && err != errors.ErrNotFound {
```

Comment:
so that the error can be wrapped with context

```suggestion
	if err != nil && !errors.Is(err, errors.ErrNotFound) {
```